### PR TITLE
Thai/account creation

### PR DESCRIPTION
**Pull request by athai**

### CODE REVIEW

**Comment by tschuy** [comment]
Can you add documentation on how to use this?

**Comment by tschuy** [comment]
I don't particularly want my password to be printed to the screen when I type it in. does prompt.js have a facility to hide this?

It should handle duplicate users cleanly instead of printing an Unhandled rejection Error.

**Comment by athai** [comment]
Yeah, the password can be hidden during the prompts.  Do you not want the password to print to the screen after creating the account either?

Also, what would be added to the documentation other than:
run "npm run create-account" to start the script

**Comment by tschuy** [comment]
Passwords should never be displayed, ever.

And mostly that, maybe with a nice sentence saying "there's a script included to create users".

**Comment by tschuy** [comment]
Can you rebase or merge develop? And the script should probably go in ``scripts/``, not ``src/``.

File: scripts/create-account.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env node
+
+//To start the script, run 'npm run create-account'
+
+//Library requirements
```
Comment (tschuy) [inline scripts/create-account.js]: One space after //s before writing comments.

File: src/create-account.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env node
```
Comment (tschuy) [inline src/create-account.js]: Remove this file

File: scripts/create-account.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                })
+            //Exits the process after storing user info in the db
+            .then(
+                function() {
+                    process.exit(0);
+                });
+        });
+    });
+
+    console.log('\nCommand-line input received:\n');
```
Comment (tschuy) [inline scripts/create-account.js]: Can you just get rid of this section entirely? I feel like printing the username isn't particularly useful, especially because it's shown five lines above where i typed it.

File: scripts/create-account.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+            .then(
+                function() {
+                    console.log('\nUser successfully created\n');
+                })
+            /*If the user enters a duplicate name, information will not be
+              added to the database. An error message will print to the screen
+              and the process will exit. */
+            .catch(
+                function() {
+                    console.log('\nINVALID ENTRY: That username is already' +
```
Comment (tschuy) [inline scripts/create-account.js]: We'll want to make sure to separate "I don't know what went wrong" from "the username is already taken".

Comment (tschuy) [inline scripts/create-account.js]: Knex passes an error variable to the catch function. If you add it as a paramter to your function, you can check the result code of the error (err.errno) against SQLite's Result Codes list:

https://www.sqlite.org/c3ref/c_abort.html

Comment (tschuy) [inline scripts/create-account.js]: Though I'd also add a note saying that this needs to be fixed to work on databases other than sqlite.

File: scripts/create-account.js [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+                    // to work on databases other than sqlite
+
+                    /* If the error that occurs is not a constraint violation,
+                       an error message will print to the screen with a link to
+                       a list of result codes. The errno is also printed to the
+                       screen. */
+                    else {
+                        console.log('Something went wrong! Check out ' +
+                                    'https://www.sqlite.org/c3ref/' +
+                                    'c_abort.html to figure out what' +
```
Comment (ElijahCaine) [inline scripts/create-account.js]: Add a space after ``what``.
```
$ npm run create-account

> timesync@0.0.0 create-account /path/to/timesync-node
> node ./scripts/create-account.js

prompt: name:  user
prompt: password:
Something went wrong! Check out https://www.sqlite.org/c3ref/c_abort.html to figure out whathappened.
  Your error number is:
1
```

File: src/create-account.js (hunk 2 above)
Comment (iankronquist) [inline src/create-account.js]: In the future you should use `git mv src dest`. It moves and adds the file.

File: docs/source/dev/usage.rst [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+Usage Guide
+===========
+
+Creating a TimeSync Account
+---------------------------
+
+The TimeSync API includes a handy little script to create users. To start it
+up, run::
+
+    npm run create-account
```
Comment (ElijahCaine) [inline docs/source/dev/usage.rst]: We should add a little section about if you get this error:
```
$ npm run create-account
                                                                                                          
> timesync@0.0.0 create-account /path/to/timesync-node
> node ./scripts/create-account.js

prompt: name:  username
prompt: password:
Something went wrong! Check out https://www.sqlite.org/c3ref/c_abort.html to figure out what happened.

  Your error number is:
1
```
You probably need to run
```
$ npm run migrations
```
that solved the problem for me anyway.

Really some docs on expected output and how to fix basic problems would be nice.

File: docs/source/dev/usage.rst [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+up, run::
+
+    $ npm run create-account
+
+This will prompt for a username and a password and store the information into
+the database.
+
+
+Expected Outputs
+----------------
```
Comment (ElijahCaine) [inline docs/source/dev/usage.rst]: This should be a subsection of the ``Creating a TimeSync Account``. This is done by replacing the ``---`` underlining with ``~~~``s.

File: docs/source/dev/usage.rst [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+    > timesync@0.0.0 create-account /home/thai/projects/timesync-node
+    > node ./scripts/create-account.js
+
+    prompt: name:  username
+    prompt: password:  
+
+    User successfully created
+
+However if a username is unavailable, you'll be met with::
```
Comment (ElijahCaine) [inline docs/source/dev/usage.rst]: I would separate the errors into their own subsection. This makes searching and linking easier.

File: docs/source/dev/usage.rst [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    prompt: password:  
+
+    Something went wrong! Check out https://www.sqlite.org/c3ref/c_abort.html to figure out what happened.
+
+    Your error number is: 1
+
+    Exiting...
+
+According to SQlite's `result code definitions`_, an errno of 1 is indicative
+of either a SQL error or missing database. Odds are (with this script), it's
```
Comment (tschuy) [inline docs/source/dev/usage.rst]: "odds are, if you're running this in a development environment, it's the latter."

File: scripts/create-account.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+var db = process.env.DATABASE || 'development';
+
+// Load the database (default = development)
+var knex = require('knex')(knexfile[db]);
+
+// Prompt user for their information
+var info = {
+    properties: {
+        name: {
+            // Note to self: '\w' matches all alphanumeric char and '_'
```
Comment (tschuy) [inline scripts/create-account.js]: Note, not note to self.

**Comment by tschuy** [comment]
Other than that, it looks good for a development account creation tool. We'll definitely want to generalize it in the future, though.

**Comment by athai** [comment]
@tschuy

**Comment by iankronquist** [comment]
You need to add `#!/usr/bin/env node` if you're going to make the script executable.

```
(timesync)iankronquist@puppettop:(timesync)(thai/account-creation) → ./scripts/create-account.js 
./scripts/create-account.js: line 1: //: is a directory
./scripts/create-account.js: line 3: //: is a directory
./scripts/create-account.js: line 4: syntax error near unexpected token `('
./scripts/create-account.js: line 4: `var prompt = require('prompt');'
```
I put `test` for both the username and password, and this is what I got. I forgot to run migrations. Maybe you should remind me to delete the database and run migrations.
```
node ./scripts/create-account.js
prompt: name:  test
prompt: password:  

Something went wrong! Check out https://www.sqlite.org/c3ref/c_abort.html to figure out what happened.

  Your error number is: 1

  Exiting...


```
I think you should print usage information if someone runs `node ./scripts/create-account.js --help` or `node ./scripts/create-account.js -h`.

Good work!

**Comment by tschuy** [comment]
@iankronquist read the usage docs :)

File: scripts/create-account.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+                                    'https://www.sqlite.org/c3ref/' +
+                                    'c_abort.html to figure out what ' +
+                                    'happened.\n');
+                        console.log('  Your error number is: ' + err.errno);
+                        console.log('\n  Exiting...\n');
+                    }
+                })
+            // Exits the process after storing user info in the db
+            .then(
+                function() {
```
Comment (iankronquist) [inline scripts/create-account.js]: Functions which take no arguments and just call another function are not necessary. Delete lines 78 and 80.

Comment (tschuy) [inline scripts/create-account.js]: Yes they are. They're only useless if you don't need to pass anything to the function they wrap.

In this instance, try calling ``process.exit()`` with no params. If that works, **then** lines 78 and 80 are unnecessary.

Comment (iankronquist) [inline scripts/create-account.js]: This works just fine for me: `.then( process.exit(0)); `

Comment (athai) [inline scripts/create-account.js]: Deleting lines 78 and 80 makes the error catching fail. It also doesn't store the newly created account into the db correctly. Ditto for line 43.

Comment (tschuy) [inline scripts/create-account.js]: I think you should probably just leave them.

**Comment by tschuy** [comment]
I don't think those comments are necessary.

**Comment by athai** [comment]
The ones about the functions and why they're there?

**Comment by tschuy** [comment]
Yeah. That's just standard fare for a promise, there's no extra explanation (other than a good blog post explaining promises) needed.

**Comment by ElijahCaine** [comment]
@athai Everything here looks good. Your examples in the docs are very good and the behavior described is accurate to experience. It gets the job done.

We may need to add logic like ensuring the user has entered a secure password (we can create a user with password ``1`` right now) but that can be decided on later as we have to make design choices before that code can be written.

tldr: Looks good to me. +1

**Comment by tschuy** [comment]
Looks good to me too :) +1

**Comment by Kennric** [comment]
+1
